packages/mobile/src/geth/saga.ts

```diff
@@ -210,29 +211,38 @@ export const _waitForGethInit = waitForGethInit
 export function* initGethSaga() {
   Logger.debug(TAG, 'Initializing Geth')
   yield put(setInitState(InitializationState.INITIALIZING))
+  let failureResult: GethInitOutcomes
 
-  const { result } = yield race({
-    result: call(waitForGethInit),
-    timeout: delay(INIT_GETH_TIMEOUT),
-  })
+  while (true) {
+    const { result } = yield race({
+      result: call(waitForGethInit),
+      timeout: delay(INIT_GETH_TIMEOUT),
+    })
 
-  let restartAppAutomatically: boolean = false
-  let errorContext: string
-  switch (result) {
-    case GethInitOutcomes.SUCCESS: {
+    if (result === GethInitOutcomes.SUCCESS) {
       Logger.debug(TAG, 'Geth initialized')
       ValoraAnalytics.track(GethEvents.geth_init_success)
       yield put(setInitState(InitializationState.INITIALIZED))
       return
-    }
-    case GethInitOutcomes.NETWORK_ERROR_FETCHING_STATIC_NODES: {
-      errorContext =
+    } else if (result === GethInitOutcomes.NETWORK_ERROR_FETCHING_STATIC_NODES) {
+      // TODO(erdal): we might want to retry in other error cases as well
+      // analyze stats and decide
+      Logger.error(
+        TAG,
         'Could not fetch static nodes from the network. Tell user to check data connection.'
-      Logger.error(TAG, errorContext)
+      )
       yield put(setInitState(InitializationState.DATA_CONNECTION_MISSING_ERROR))
-      restartAppAutomatically = false
+      yield delay(GETH_RETRY_DELAY)
+    } else {
+      failureResult = result
       break
     }
+  }
+
+  let restartAppAutomatically: boolean = false
+  let errorContext: string
+
+  switch (failureResult) {
     case GethInitOutcomes.NETWORK_ERROR_FETCHING_GENESIS_BLOCK: {
```

Comment:
Looks like we'd have a similar issue when fetching the genesis block on the first app launch if the network is down.
Right?

Reply:
Yes, exactly! also the other remaining case is geth timeout. I wanted to discuss with you possible ways to recover from those. I tried using the same code path for those errors as well but we run into multiple geth inits happening at the same time. We need a clean way to cancel before we init a new one.

Or, if geth itself has any sort of timeouts, we need to make our own timeouts at least as large as those, so that when we retry we know that the previous one stopped trying. Does that make sense? I would love to hear your thoughts and what we've already tried in the past.

Reply:
Yes let's continue the discussion. Can you open an issue so we can address these edge cases in a separate PR? 🙏